lib/concurrent-ruby/concurrent/atomic/locals.rb

```diff
@@ -0,0 +1,184 @@
+require 'concurrent/constants'
+
+module Concurrent
+  # @!visibility private
+  # @!macro internal_implementation_note
+  #
+  # An abstract implementation of local storage, with sub-classes for
+  # per-thread and per-fiber locals.
+  #
+  # Each execution context (EC, thread or fiber) has a lazily initialized array
+  # of local variable values. Each time a new local variable is created, we
+  # allocate an "index" for it.
+  #
+  # For example, if the allocated index is 1, that means slot #1 in EVERY EC's
+  # locals array will be used for the value of that variable.
+  #
+  # The good thing about using a per-EC structure to hold values, rather than
+  # a global, is that no synchronization is needed when reading and writing
+  # those values (since the structure is only ever accessed by a single
+  # thread).
+  #
+  # Of course, when a local variable is GC'd, 1) we need to recover its index
+  # for use by other new local variables (otherwise the locals arrays could
+  # get bigger and bigger with time), and 2) we need to null out all the
+  # references held in the now-unused slots (both to avoid blocking GC of those
+  # objects, and also to prevent "stale" values from being passed on to a new
+  # local when the index is reused).
+  #
+  # Because we need to null out freed slots, we need to keep references to
+  # ALL the locals arrays, so we can null out the appropriate slots in all of
+  # them. This is why we need to use a finalizer to clean up the locals array
+  # when the EC goes out of scope.
+  class AbstractLocals
+    def initialize(name_prefix = :concurrent_locals)
+      @free = []
+      @lock = Mutex.new
+      @all_locals = {}
+      @next = 0
+
+      @name = :"#{name_prefix}_#{object_id}"
+    end
+
+    def synchronize
+      @lock.synchronize { yield }
+    end
+
+    if Concurrent.on_cruby?
+      def weak_synchronize
+        yield
+      end
+    else
+      alias_method :weak_synchronize, :synchronize
+    end
+
+    def next_index(target)
+      index = synchronize do
+        if @free.empty?
+          @next += 1
+        else
+          @free.pop
+        end
+      end
+
+      # When the target goes out of scope, we should free the associated index
+      # and all values stored into it.
+      ObjectSpace.define_finalizer(target, target_finalizer(index))
+
+      return index
+    end
+
+    def free_index(index)
+      weak_synchronize do
+        # This is O(N) number of ECs with locals... I'm concerned about the
+        # performance cost. But I don't see a better way to do it. Using
+        # `each_value` can blow up if @all_locals is modified during iteration.
+        # Maybe using a linked list for `@all_locals` would be better?
```

Comment:
The previous comment here had more information:
```
          # The cost of GC'ing a TLV is linear in the number of threads using TLVs
          # But that is natural! More threads means more storage is used per TLV
          # So naturally more CPU time is required to free more storage
          #
          # DO NOT use each_value which might conflict with new pair assignment
          # into the hash in #value= method
```
Could you add it back on top or below your comment?

Reply:
I'll remove my comment as they basically say the same thing. However, I fundamentally dislike:

> So naturally more CPU time is required to free more storage.

Essentially, freeing the storage is O(number of threads/fibers). On a system with lots of instances, that will be slow. However, it should be an infrequent operation. I personally think we should be aiming for constant time if possible.

Reply:
I don't think constant time is possible here, which is the case of a ThreadLocalVar/FiberLocalVar being GC'd. We need to clean (up to) N memory locations, one for each Thread/Fiber which used it (and since this is globalized, one for each Thread/Fiber which ever used a ThreadLocalVar/FiberLocalVar).

Reply:
Using `@all_locals.values` is also a potentially expensive memory allocation, and O(2N) because it will iterate the list twice (once for the initial copy into the temporary array, and once for the actual iteration). Well, I'm not trying to solve that problem in this PR, but I don't think it's good for scalability.